espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ComposeNodeFinder.kt

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.appium.espressoserver.lib.helpers
+
+import androidx.compose.ui.test.*
+import io.appium.espressoserver.lib.handlers.exceptions.InvalidSelectorException
+import io.appium.espressoserver.lib.model.Strategy
+import io.appium.espressoserver.EspressoServerRunnerTest
+import io.appium.espressoserver.lib.handlers.exceptions.AppiumException
+import io.appium.espressoserver.lib.handlers.exceptions.InvalidArgumentException
+import io.appium.espressoserver.lib.handlers.exceptions.StaleElementException
+import io.appium.espressoserver.lib.model.Locator
+import java.util.NoSuchElementException
+
+/**
+ * Retrieve cached node and return the SemanticsNodeInteraction
+ */
+fun getNodeInteractionById(elementId: String?): SemanticsNodeInteraction =
+    elementId?.let { ComposeViewCache.get(it) ?: throw StaleElementException(it) }
+        ?: throw InvalidArgumentException("Cannot find 'null' element")
+
+
+/**
+ * Retrieve cached SemanticsNodeInteraction and ensure its not stale
+ */
+fun getValidNodeInteractionById(elementId: String?): SemanticsNodeInteraction =
+    try {
+        val nodeInteraction = getNodeInteractionById(elementId)
+        nodeInteraction.assertExists().let { return nodeInteraction }
```

Comment:
basically here we perform two requests to the UI framework: the first one to check if the element exists and the second one, to, for example, click the element. I was wondering if we could prevent this by catching an appropriate exception on `click` and skipping `assertExists` check, so then only one request is executed. For example:

```
val nodeInteraction = getNodeInteractionById(elementId)
try {
  nodeInteraction.performClick()
} catch (e: ElementDoesNotExistComposeException) {
  throw StaleElement
} catch (e: ElementCannotBeClickedComposeException) {
  throw IllagalElementState
}
```

Reply:
All the perform actions (eg `SemanticsNodeInteraction.performClick`) calls  `SemanticsNodeInteraction#fetchSemanticsNode` which throws `AssertionError`. I wanted to avoid catching exeptions in every individual action from our side later, and therefore used `getValidNodeInteractionById `. Instead of `assertExists` we can possibly just `fetchSemanticsNode` and catch `AssertionError` like
```
fun getValidNodeInteractionById(elementId: String?): SemanticsNodeInteraction =
    try {
        val nodeInteraction = getNodeInteractionById(elementId)
        nodeInteraction.fetchSemanticsNode()
        nodeInteraction
    } catch (e: AssertionError) {
        throw StaleElementException(elementId!!)
    }
```

Reply:
this won't be different from the current implementation, because `nodeInteraction.fetchSemanticsNode()` also requests the UI

Reply:
Makes sense, added exception handling in click